openrr-gui/src/velocity_sender.rs

```diff
@@ -1,683 +1,242 @@
-    other: Option<String>,
-    update_on_error: bool,
-}
-
-impl Errors {
-    fn is_none(&self) -> bool {
-        self.velocity_state.is_none() && self.other.is_none()
-    }
-
-    fn skip_update(&mut self, message: &Message) -> bool {
-        self.update_on_error = false;
-        // update always if there is no error.
-        if self.is_none() {
-            return false;
-        }
-
-        if self.velocity_state.is_none() && self.other.is_some() {
-            self.other = None;
-            // Other errors are usually unresolvable by user's action.
-            return false;
-        }
-
-        match message {
-            Message::SliderChanged { index, .. }
-            | Message::SliderTextInputChanged { index, .. }
-                if self.velocity_state.is_some()
-                    && self.velocity_state.as_ref().unwrap().0 == *index =>
-            {
-                false
-            }
-            _ => {
-                self.update_on_error = true;
-                true
-            }
-        }
-    }
-}
-
-struct VelocitySender<M>
+pub struct VelocitySender<M>
 where
     M: MoveBase + 'static,
 {
     move_base: M,
 
-    /// x, y, theta
-    velocity_state: [SliderState; 3],
+    velocity: BaseVelocity,
     show_velocity: bool,
-    /// x+
-    up_button: ButtonState,
-    /// x-
-    down_button: ButtonState,
-    /// y+
-    left_button: ButtonState,
-    /// y-
-    right_button: ButtonState,
-    /// theta+
-    counterclockwise_button: ButtonState,
-    /// theta-
-    clockwise_button: ButtonState,
-    /// set velocity to 0.
-    stop_button: button::State,
 
-    errors: Errors,
-    theme: Theme,
+    up_button: RetainedImage,
+    down_button: RetainedImage,
+    left_button: RetainedImage,
+    right_button: RetainedImage,
+    counterclockwise_button: RetainedImage,
+    clockwise_button: RetainedImage,
 }
 
 impl<M> VelocitySender<M>
 where
     M: MoveBase + 'static,
 {
-    fn new(move_base: M, theme: Theme) -> Self {
-        // https://github.com/hecrj/iced/issues/846
-        #[cfg(feature = "glow")]
-        warn!("glow backend is not fully supported");
+    fn new(move_base: M) -> Self {
+        macro_rules! button_image {
+            ($alt:expr, $img_path:expr $(,)?) => {
+                RetainedImage::from_image_bytes(
+                    $alt,
+                    include_bytes!(concat!("../assets/material-design-icons/", $img_path)),
+                )
+                .unwrap()
+            };
+        }
 
         Self {
             move_base,
-            velocity_state: [
-                SliderState::new("x", -1.0..=1.0),
-                SliderState::new("y", -1.0..=1.0),
-                SliderState::new("θ", -1.0..=1.0),
-            ],
+            velocity: BaseVelocity {
+                x: 0.0,
+                y: 0.0,
+                theta: 0.0,
+            },
+            up_button: button_image!("↑", "baseline_arrow_upward_white_24dp.png"),
```

Comment:
I want to know how we get the png files.
Is it embedded in egui?

Reply:
They are included in [openrr-gui/assets/material-design-icons](https://github.com/openrr/openrr/tree/taiki-e/main/openrr-gui/assets/material-design-icons).